File: 06-milking-cows/milk2.cpp
/*
ID: caleb.h1
LANG: C++
PROG: milk2
*/

#include <fstream>
#include <map>

#define MAX(a, b) a > b ? a : b;
#define MIN(a, b) a < b ? a : b;

int main() {
    std::ifstream infile("milk2.in");

    // Beginning and ending times 1->5,000
    unsigned int numTimes;
    infile >> numTimes;

    // Maps starting times to their ending times
    // Ordered!
    std::map<uint, uint> times;

    uint s, e;
    for (unsigned int i=0; i<numTimes; ++i) {
        infile >> s >> e;
        // Duplciate checking
        try  {
            times[s] = MAX(times[s], e);
        }
        catch (int e) {
            times[s] = e;
        }
    }

    infile.close();


    // merge
    std::map<uint, uint>::iterator next;
    for (std::map<uint, uint>::iterator i = times.begin(); i != times.end(); ++i) {
        while (std::next(i) != times.end() && std::next(i)->first <= i->second) { // Don't go too far && next.start <= cur.end
            i->second = MAX(std::next(i)->second, i->second);
            times.erase(std::next(i));
        }
    }

    // Find longest duration of milking
    // Find longest duration without milking
    uint durationMilking = 0;
    uint durationNoMilking = 0;

    for (std::map<uint, uint>::iterator i = times.begin(); i != times.end(); ++i) {
        durationMilking = MAX(durationMilking, i->second - i->first);
        durationNoMilking = MAX(durationNoMilking, (std::next(i) != times.end() ? std::next(i)->first - i->second : 0));
    }


    std::ofstream outfile("milk2.out");
    outfile << durationMilking << ' ' << durationNoMilking << std::endl;
    outfile.close();
}

File: 06-milking-cows/milk2.py
"""
ID: caleb.h1
LANG: PYTHON3
PROG: milk2
"""


def main():
    with open('milk2.in', 'r') as f:

        # Beginning and ending times 1->5,000
        numTimes = int(f.readline())

        # Maps starting times to their ending times
        times = dict()

        for i in range(numTimes):
            s, e = (int(n) for n in f.readline().strip().split())
            # Duplciate checking
            try:
                times[s] = max(times[s], e)
            except:
                times[s] = e
        
    times = dict(sorted(times.items()))

    times = [[key, times[key]] for key in times]

    # merge


    i = 0
    while i < len(times):
        while i + 1 < len(times) and times[i+1][0] <= times[i][1]:
            times[i][1] = max(times[i][1], times[i+1][1]) # cur.end = max(cur.end, next.end)
            times.pop(i+1) # merge times together by removing overlapping one
        i+=1
    
    milkingDuration = max( [ i[1] - i[0] for i in times ] )
    noMilkingDuration = 0
    for i in range(len(times) - 1):
        noMilkingDuration = max(times[i+1][0] - times[i][1], noMilkingDuration)
    
    with open('milk2.out', 'w') as f:
        f.write(f'{milkingDuration} {noMilkingDuration}\n')



if __name__ == '__main__':
    main()